src/NLog/Layouts/XmlLayout.cs

```diff
@@ -0,0 +1,504 @@
+        public string PropertiesNodeValueAttribute { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmlLayout"/> class.
+        /// </summary>
+        public XmlLayout()
+            :this("logevent", null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="XmlLayout"/> class.
+        /// </summary>
+        /// <param name="nodeName">The name of the top XML node</param>
+        /// <param name="nodeValue">The value of the top XML node</param>
+        public XmlLayout(string nodeName, Layout nodeValue)
+        {
+            NodeName = nodeName;
+            NodeValue = nodeValue;
+            Attributes = new List<XmlAttribute>();
+            Nodes = new List<XmlLayout>();
+            ExcludeProperties = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Initializes the layout.
+        /// </summary>
+        protected override void InitializeLayout()
+        {
+            base.InitializeLayout();
+            if (IncludeMdc)
+            {
+                ThreadAgnostic = false;
+            }
+#if !SILVERLIGHT
+            if (IncludeMdlc)
+            {
+                ThreadAgnostic = false;
+            }
+#endif
+            if (Attributes.Count > 1)
+            {
+                HashSet<string> attributeValidator = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                foreach (var attribute in Attributes)
+                {
+                    if (string.IsNullOrEmpty(attribute.Name))
+                    {
+                        Common.InternalLogger.Warn("XmlLayout(NodeName={0}): Contains attribute with missing name (Ignored)");
+                    }
+                    else if (attributeValidator.Contains(attribute.Name))
+                    {
+                        Common.InternalLogger.Warn("XmlLayout(NodeName={0}): Contains duplicate attribute name: {1} (Invalid xml)", NodeName, attribute.Name);
+                    }
+                    else
+                    {
+                        attributeValidator.Add(attribute.Name);
+                    }
+                }
+            }
+        }
+
+        internal override void PrecalculateBuilder(LogEventInfo logEvent, StringBuilder target)
+        {
+            if (!ThreadAgnostic) RenderAppendBuilder(logEvent, target, true);
+        }
+
+        /// <summary>
+        /// Formats the log event as a XML document for writing.
+        /// </summary>
+        /// <param name="logEvent">The logging event.</param>
+        /// <param name="target"><see cref="StringBuilder"/> for the result</param>
+        protected override void RenderFormattedMessage(LogEventInfo logEvent, StringBuilder target)
+        {
+            int orgLength = target.Length;
+            RenderXmlFormattedMessage(logEvent, target);
+            if (target.Length == orgLength && IncludeEmptyValue && !string.IsNullOrEmpty(NodeName))
```

Comment:
if `IncludeEmptyValue == true`, should't the whole `if` result in `true` if `nodeName == ""`?

Reply:
Not sure I understand. It outputs the NodeName, even if nothing to output. If no NodeName, what should it then output?